packages/language-server/src/format/comments.ts

```diff
@@ -0,0 +1,113 @@
+            const commentToken = comment.node;
+            if (commentToken.trailing || commentToken.leading) {
+                return;
+            }
+            const commentDoc = printComment(commentToken, options);
+            if (commentToken.tokenType.name === TokenType.MULTI_LINE_COMMENT) {
+                parts.push(commentDoc);
+            } else {
+                parts.push(lineSuffix(commentDoc));
+            }
+        }, "comments");
+        return [join(hardline, parts)];
+    } else {
+        return [];
+    }
+}
+/**
+ * Prints a comment
+ *
+ * @param comment the comment to print
+ * @param options the formatting options
+ * @returns the formatted comment
+ */
+
+export function printComment(comment: Comment, options: Options): Doc {
+    comment.printed = true;
+    if (comment.tokenType.name === TokenType.MULTI_LINE_COMMENT) {
+        const image = comment.image;
+        const leadingWhitespaceLength = calculateWhitespaceLength(
+            getLeadingWhitespace(options.originalText, comment.startOffset),
+            options.tabWidth
+        );
+        const lines = image
+            .split("\n")
+            .map((line) => decreaseIdentation(line, leadingWhitespaceLength, options.tabWidth));
+        return join(hardline, lines);
+    } else {
+        return comment.image;
+    }
+}
+/**
+ * Gets the leading whitespace of a given offset until the last non-whitespace character or newline
+ *
+ * @param text the text to get the leading whitespace from
+ * @param offset the offset to start from
+ * @returns the leading whitespace
+ */
+function getLeadingWhitespace(text: string, offset: number): string {
+    let start = offset;
+    while (start > 0 && text[start - 1].match(/[^\S\n]/)) {
+        start--;
+    }
+    return text.substring(start, offset);
+}
+/**
+ * Calculates the whitespace length of a given text
+ *
+ * @param text the purely whitespace text
+ * @param tabWidth the width of a tab
+ * @returns the length of the text
+ */
+function calculateWhitespaceLength(text: string, tabWidth: number): number {
+    let length = 0;
+    for (let i = 0; i < text.length; i++) {
+        if (text[i] === "\t") {
+            length += tabWidth;
+        } else {
+            length++;
+        }
+    }
+    return length;
+}
+/**
+ * Decreases the indentation of a given text by a given length
+ * If a tab causes a larger dedentation than desired, spaces are used to replace it
+ *
+ * @param text the text to dedent
+ * @param length the length to dedent
+ * @param tabWidth the width of a tab
+ * @returns the dedented text
+ */
+function decreaseIdentation(text: string, length: number, tabWidth: number): string {
+    let i = 0;
+    while (length > 0 && i < text.length && text[i].match(/[^\S\n]/)) {
+        if (text[i] === "\t") {
+            length -= tabWidth;
+        } else {
+            length--;
+        }
+        i++;
+    }
+    const leadingSpaces = " ".repeat(Math.max(-length, 0));
```

Comment:
Don't you mean
```suggestion
    const leadingSpaces = " ".repeat(Math.max(length, 0));
```
?
`length` seems to be positive.

Reply:
(I mean, there is a case where `length == -1), but I thought that's what the `max(…, 0)` is for.

Reply:
length can be both positive and negative
in case it is positive, there is remaining length which could not be removed because it is no longer whitespace
e.g. you have 4 leading spaces, but want to remove 8. In this case, you just remove 4 and live with it.
However, it can also be negative. Imagine you want to remove 1 space, however, it encounters a tab character (with a tab width of 4). In this case, it removes the tab, however, length (1) then gets reduced by 4, so you end up with -3. This means, that you removed 3 spaces "too much", so you need to re-add these 3 spaces